src/Microsoft.Extensions.ML/ServiceCollectionExtensions.cs

```diff
@@ -27,19 +28,60 @@ public static class ServiceCollectionExtensions
+            return services.AddPredictionEnginePool<TData, TPrediction>(() =>
+                services.AddSingleton<PredictionEnginePool<TData, TPrediction>, PredictionEnginePool<TData, TPrediction>>());
         }
 
-        internal static IServiceCollection AddPredictionEngineServices<TData, TPrediction>(
-            this IServiceCollection services)
+        /// <summary>
+        /// Adds a <see cref="PredictionEnginePool{TData, TPrediction}"/> and required config services to the service collection.
+        /// </summary>
+        /// <param name="services">
+        /// The <see cref="IServiceCollection "/> to add services to.
+        /// </param>
+        /// <param name="implementationFactory">
+        /// The factory that creates the <see cref="PredictionEnginePoolBuilder{TData, TPrediction}"/>.
+        /// </param>
+        /// <returns>
+        /// The <see cref="PredictionEnginePoolBuilder{TData, TPrediction}"/> that was added to the collection.
+        /// </returns>
+        public static PredictionEnginePoolBuilder<TData, TPrediction> AddPredictionEnginePool<TData, TPrediction>(
+            this IServiceCollection services,
+            Func<IServiceProvider, PredictionEnginePool<TData, TPrediction>> implementationFactory)
+            where TData : class
+            where TPrediction : class, new()
+        {
+            return services.AddPredictionEnginePool<TData, TPrediction>(() =>
+                services.AddSingleton<PredictionEnginePool<TData, TPrediction>, PredictionEnginePool<TData, TPrediction>>(implementationFactory));
+        }
+
+        private static PredictionEnginePoolBuilder<TData, TPrediction> AddPredictionEnginePool<TData, TPrediction>(
+            this IServiceCollection services,
+            Action callback)
             where TData : class
             where TPrediction : class, new()
         {
-            services.AddLogging();
-            services.AddOptions();
-            services.TryAddEnumerable(ServiceDescriptor.Singleton<IPostConfigureOptions<MLOptions>, PostMLContextOptionsConfiguration>());
-            services.AddSingleton<PredictionEnginePool<TData, TPrediction>, PredictionEnginePool<TData, TPrediction>>();
+            services.AddPredictionEnginePoolConfigOnly();
+            callback();
+
+            return new PredictionEnginePoolBuilder<TData, TPrediction>(services);
+        }
+
+        /// <summary>
+        /// Adds only the required config services for <see cref="PredictionEnginePool{TData, TPrediction}"/> to the service collection.
+        /// </summary>
+        /// <param name="services">
+        /// The <see cref="IServiceCollection "/> to add services to.
+        /// </param>
+        /// <returns>
+        /// A reference to this instance after the operation has completed.
+        /// </returns>
+        public static IServiceCollection AddPredictionEnginePoolConfigOnly(this IServiceCollection services)
```

Comment:
I'm not certain `ConfigOnly` is a great name here. How about `AddRequiredPredictionEnginePoolServices`?

Reply:
or

`AddPredictionEnginePoolRequiredServices`

Reply:
The _only issue_ (so to speak) with those is the same as with `AddPredictionEnginePoolPrerequisites` in that they imply to the user that they must call this method _before_ calling `AddPredictionEnginePool`.  Did you see [my reply](https://github.com/dotnet/machinelearning/pull/4393#discussion_r344935260) to our existing conversation regarding this, particularly the reasoning behind "ConfigOnly"?

Reply:
I wouldn't use the terms "Add...Config" because it sounds like it has to do with [`Microsoft.Extensions.Configuration`](https://www.nuget.org/packages/Microsoft.Extensions.Configuration), but it doesn't.

How about `ConfigurePredictionEnginePoolServices`? To me it doesn't imply that it must be called first. With good XML docs and good samples, it will be fairly obvious to users that they can just call `AddPredictionEnginePool`, which follows the same pattern as other `Microsoft.Extensions` APIs.

Reply:
I think keeping the beginning part `Add` is better as it does indeed add services that are also added with `Add` methods. So, based on your reasoning about good XML and good samples, then I would say your original idea `AddPredictionEnginePoolPrerequisites`, or your second idea `AddRequiredPredictionEnginePoolServices` are the best options. If you're happy with one of those, let me know, and I'll make the change.

Reply:
I'd be happy with either of those names as well.

If I was forced to pick between them, I would shy away from the term "Prerequisites". It doesn't appear we use that term often in .NET APIs. So I guess I would pick `AddRequiredPredictionEnginePoolServices`

Reply:
OK, I'll go with that one. However, before I do, I need to ask your opinion on something first. I originally came up with this use case because I needed to create multiple `PredictionEnginePool`s, one for a Regression model, and one for a Time Series model. However, I learned from your colleague when he [replied here](https://github.com/dotnet/machinelearning-samples/issues/739#issuecomment-552004422) (also mentioning you) to an issue I raised that `PredictionEnginePool` cannot be used for Time Series model. Therefore, in my case, I only need one `PredictionEnginePool` for now. That's not to say the use case is void as I guess multiple `PredictionEnginePool`s would still be useful for different models that use `PredictionEnginePool`, or even the same model (Regression for instance) where the prediction (output) POCO is different. If you agree, then yes, this change is worthwhile. Let me know please, then I'll make the change.

Reply:
I _think_ this change is still valid, even if you just have 1 `PredictionEnginePool`. In your case you are using a `LazyService<PredictionEnginePool...` service, right? Even though you only need 1 PredictionEnginePool service, but you aren't calling through our normal `AddPredictionEnginePool` method to create it, you'd still want these "Required Services" added. Having to call through `AddPredictionEnginePool` would just add an unnecessary service that you aren't going to use.

Reply:
Exactly right, although the reason I was using `LazyService<PredictionEnginePool...` was because I was using multiple `PredictionEnginePool`s and only wanted one to be fully instantiated (when required). Now I'm only using one (because Time Series doesn't use `PredictionEnginePool`), it's not as crucial for it to be instantiated lazily because it will always definitely be used at some point after startup.

But the extra control is definitely useful for anyone wanting to instantiate lazily for whatever reason, or if they want additional control on adding the `PredictionEnginePool` for some other reason (maybe `AddScoped` instead of `AddSingleton`).

So I'll go ahead and make the change to the method name as discussed. :)